src/general/semidiscretization.jl

```diff
@@ -375,6 +375,29 @@ end
                     (StaticInt(v_nvariables(system)), n_moving_particles(system)))
 end
 
+# For non-`Array`s (typically GPU arrays), just reshape. Calling the `PtrArray` code above
+# for a `CuArray` yields another `CuArray` (instead of a `PtrArray`) and is 8 times slower
+# with double the allocations.
+@inline function wrap_u(u_ode, system, semi)
+    (; ranges_u) = semi
+
+    range = ranges_u[system_indices(system, semi)]
+
+    @boundscheck @assert length(range) == u_nvariables(system) * n_moving_particles(system)
+
+    return reshape(view(u_ode, range), (u_nvariables(system), n_moving_particles(system)))
+end
+
+@inline function wrap_v(v_ode, system, semi)
+    (; ranges_v) = semi
+
+    range = ranges_v[system_indices(system, semi)]
+
+    @boundscheck @assert length(range) == v_nvariables(system) * n_moving_particles(system)
+
+    return reshape(view(v_ode, range), (v_nvariables(system), n_moving_particles(system)))
+end
+
```

Comment:
How about `return wrap_vector` in `wrap_u` and `wrap_v` and dispatch for `Array`s and non-`Array`s. 
`wrap_vector` doesn't care about `u_ode` or `v_ode`.